apps/hash-graph/lib/authorization/src/backend/spicedb.rs

```diff
@@ -0,0 +1,303 @@
+            Self::Request(error) => fmt::Display::fmt(error, fmt),
+            Self::Rpc(status) => write!(fmt, "Error {}: {}", status.code, status.message),
+        }
+    }
+}
+
+impl Error for InvocationError {}
+
+enum InvocationResult<T> {
+    Success(T),
+    Failure(RpcStatus),
+}
+
+impl SpiceDb {
+    async fn request<R: DeserializeOwned>(
+        &self,
+        path: &'static str,
+        body: &(impl Serialize + Sync),
+    ) -> Result<InvocationResult<R>, reqwest::Error> {
+        let result = self
+            .configuration
+            .client
+            .execute(
+                self.configuration
+                    .client
+                    .post(format!("{}{}", self.configuration.base_path, path))
+                    .header(
+                        "Authorization",
+                        format!("Bearer {}", self.configuration.key),
+                    )
+                    .json(&body)
+                    .build()?,
+            )
+            .await?;
+
+        if result.status().is_success() {
+            Ok(InvocationResult::Success(result.json().await?))
+        } else {
+            Ok(InvocationResult::Failure(result.json().await?))
+        }
+    }
+
+    async fn call<R: DeserializeOwned>(
+        &self,
+        path: &'static str,
+        body: &(impl Serialize + Sync),
+    ) -> Result<R, Report<InvocationError>> {
+        match self
+            .request(path, body)
+            .await
+            .map_err(InvocationError::Request)?
+        {
+            InvocationResult::Success(result) => Ok(result),
+            InvocationResult::Failure(status) => bail!(InvocationError::Rpc(status)),
+        }
+    }
+}
+
+impl AuthorizationBackend for SpiceDb {
+    async fn check<S, A, R>(
+        &self,
+        subject: &S,
+        affiliation: &A,
+        resource: &R,
+        consistency: Consistency<'_>,
+    ) -> Result<CheckResponse, Report<CheckError>>
+    where
+        S: Subject + Sync,
+        A: Affiliation<R> + Sync,
+        R: Resource + Sync,
+    {
+        #[derive(Serialize)]
+        #[serde(rename_all = "camelCase", bound = "")]
+        struct RequestBody<'a, O: Resource, R: Affiliation<O>, S: Subject> {
+            consistency: schema::Consistency<'a>,
+            resource: ResourceReference<'a, O>,
+            permission: &'a R,
+            subject: SubjectReference<'a, S>,
+        }
+
+        #[derive(Deserialize)]
+        enum Permissionship {
+            #[serde(rename = "PERMISSIONSHIP_NO_PERMISSION")]
+            NoPermission,
+            #[serde(rename = "PERMISSIONSHIP_HAS_PERMISSION")]
+            HasPermission,
+            #[serde(rename = "PERMISSIONSHIP_CONDITIONAL_PERMISSION")]
+            Conditional,
+        }
+
+        #[derive(Deserialize)]
+        #[serde(rename_all = "camelCase")]
+        struct RequestResponse {
+            checked_at: schema::ZedToken<'static>,
+            permissionship: Permissionship,
+        }
```

Comment:
specific reason why all of these were implemented in the function body instead of outside? (just a subjective style thing, I try to avoid declaring types in function bodies as it tends to decrease readability quite a bit IMO)

Reply:
The structs are only expected to be used here so the scope is not polluted. As these are only relevant here the definitions are closer to the location where they are used.